### ISSUE TITLE
Plugins cannot be loaded with Pelican 4.5.0

### ISSUE THREAD

**Issue by shniubobo** [mosra/m.css#181]
With Pelican 4.5.0, the plugins cannot be loaded with `PLUGINS += ['m.x']` (where `x` is the plugin's name). A workaround is to put plugins outside of the `m` directory and use `PLUGINS += ['x']` instead.

This is to be fixed in later versions of Pelican.

Related issues: #178, https://github.com/getpelican/pelican/issues/2805

**Comment by egberts** [comment]
Yeah,  I got the same.
Did some debugging.
Noticed that Pelican 4.5 cannot use `namespace_plugins` approach, so it resorts to `load_legacy()` approach.

**Comment by egberts** [comment]
I'm running Pelican 4.5.0 and am still experiencing load_legacy() not found issue.
```console
$ python3.8 -c "import pelican;print(pelican.__version__)"
4.5.0
```
At 7a6686f4 (origin/master, origin/HEAD) Fix SUMMARY_END_SUFFIX variable name in changelog
 

I've done a `git submodule` checkout of m.css into the working area of a website such that this directory layout is:
```
+ egbert.net
|-- content
|-- output
|-- plugins
|-- m.css
   |
   +-- plugins
      +-- ansilexer.py
      +-- dot2svg.py
      +-- latex2svg.py
      +-- latex2svgextra.py
      +-- m
         +-- htmlsanity.py
         +-- metadata.py 
...
```

I used the following plugin settings:
```
PLUGIN_PATHS = [ './plugins', 'm.css/plugins' ]
PLUGINS = [
#          'i18n_subsites',
          'collate_content',
          'sitemap',
          'dateish',
          'tag_cloud',
          'm.alias',
          'm.htmlsanity',
          'm.images',
          'm.math',
          'just_table',
          'code_include',
]
```

And my console output for Pelican was:
```console
ERROR: Cannot load plugin `m.alias`
  | No module named 'm'
ERROR: Cannot load plugin `m.htmlsanity`
  | No module named 'm'
ERROR: Cannot load plugin `m.images`
  | No module named 'm'
ERROR: Cannot load plugin `m.math`
  | No module named 'm'
```
Surely, I must be misreading the documentation or something, am I?

**Comment by shniubobo** [comment]
> Surely, I must be misreading the documentation or something, am I?

From https://github.com/mosra/m.css/issues/181#issue-718676189:
> This is **to be** fixed in **later versions** of Pelican.

**Comment by egberts** [comment]
I wonder if it would be made easier if the period in m.css is replaced with an underscore “_” symbol.

**Comment by mosra** [comment]
Nooo, that's ugly! :trollface: 

In all seriousness, there's a proposed patch, so I think it's easier for everyone involved to just stay on 4.2 until this behavior is restored. Changing everything to an underscore means I need to change all docs, break compatibility with 4.2 (which would make a lot of existing users angry) and break also Python and Doxygen themes (because these use those plugins too).

**Comment by egberts** [comment]
Unfortunately, I’ve taken the proposed fix and tried a few variants of my own but it looks like legacy plugins are screwed until next Python subsubversion.

**Comment by ikeingtonbear** [comment]
I seem to be having a related issue, but not with m.[plugin] plugins, rather the plugins directly under the `m.css/plugins` directory. I'm trying to use the code plugin, but it's unable to load the ansilexer moulde.

Here is some of my debug output:
```
DEBUG: Pelican version: 4.5.3
DEBUG: Python version: 3.9.1
DEBUG: Adding current directory to system path
DEBUG: Finding namespace plugins
DEBUG: Loading plugin `m.htmlsanity`
DEBUG: Loading plugin `m.code`
ERROR: Cannot load plugin `m.code`
  | No module named 'ansilexer'
DEBUG: Loading plugin `m.components`
DEBUG: Loading plugin `m.images`
DEBUG: Registering plugin `m.htmlsanity`
DEBUG: Registering plugin `m.components`
DEBUG: Registering plugin `m.images`
```

Snippets from my pelicanconf.py file:

```
PLUGIN_PATHS = ['m.css/plugins']
PLUGINS = ['m.htmlsanity', 'm.code', 'm.components', 'm.images']
```

If this is unrelated please let me know and I'll submit a new issue.

**Comment by benlawraus** [comment]
For m.code to work, change to
`from mcss.plugins import ansilexer`
in code.py.

**Comment by benlawraus** [comment]
Same problem on Pelican 4.7.

**Comment by shniubobo** [comment]
> Same problem on Pelican 4.7.

I cannot reproduce the problem on either pelican 4.7.0 or 4.7.1 with the methods specified at https://github.com/getpelican/pelican/issues/2805 (the second one). Could you please specify how to reproduce it on pelican 4.7?

**Comment by lpirl** [comment]
@benlawraus: I can confirm the statement of @shniubobo that plugin loading generally works again (tested Pelican 4.7.0 and 4.7.1); e.g. with:
``` python
PLUGIN_PATHS.append('themes/m.css/plugins')
PLUGINS.append('m.htmlsanity')
```
But I also did experience the behavior that @ikeingtonbear mentioned. In my case the plugin `m.dot` could not be loaded (got `No module named 'dot2svg'`). So apparently putting `themes/m.css/plugins` in `PLUGIN_PATHS` does not suffice to have this path in the `sys.path`/`PYTHONPATH`. I don't know where this should be fixed (plugin, Pelican, …?) but putting `themes/m.css/plugins` in the `sys.path` worked for me as a workaround. In your Pelican configuration, this would look like:
``` python
…
import sys
…
PLUGIN_PATHS.append('themes/m.css/plugins')
…
PLUGINS.append('m.dot')
# maybe add a comment here:
sys.path.append('themes/m.css/plugins')
```

**Comment by MinchinWeb** [comment]
I was mucking around with my Pelican plugins and ran into this same issue (I think). I was able to solve it my moving the plugin from a "raw" file into the `__init__.py` file of a folder. So `cname.py` became `cname/__init__.py`. Your call on whether that is "beautiful" enough to implement :)

**Comment by mosra** [comment]
Okay, sorry everyone that it took half a century for me to finally *act* on this matter. The remaining issue with the paths should be ~~hacked up~~ fixed in 47c005b8523437ff28001c8c47a6360f985c2a28 (@lpirl thanks for providing a simple-enough idea for a fix). I'm now running on Pelican 4.7.1 locally and my excessive use of all plugins seems to work almost as good as it did back then with 4.2.

Now onto the remaining 100s of unresolved issues...

**Comment by jegrami** [comment]
Hi, @mosra! I'm experiencing two issues I suspect are related to the m.css plugins. 

1. When I make changes to the content of my articles, they don't reflect on the served website. 
2. Special directives provided by the m.css reST plugins (`note-dim::`, `note-info::`, `note-primary::`, `gh`, etc) also doesn't reflect on the website. 

In my settings file, `CACHE_CONTENT` and `LOAD_CONTENT_CACHE` are both set to False. And the m.css plugins folder is correctly referenced there too:
```sh
PLUGIN_PATHS = ['m.css/plugins']

PLUGINS = ['m.abbr',
           'm.alias',
           'm.code',
           'm.components',
           'm.dox',
           'm.dot',
           'm.filesize',
           'm.gl',
           'm.gh',
           'm.htmlsanity',
           'm.images',
           'm.link',
           'm.math',
           'm.metadata',
           'm.plots',
           'm.sphinx',
           'm.qr',
           'm.vk']
```

The following is the Terminal output when I run `pelican -Dlr`:

```
(.venv) PS C:\Users\jegrami.de\blog> pelican -Dlr
[13:21:53] DEBUG    Pelican version: 4.9.1                                                              __init__.py:627
           DEBUG    Python version: 3.12.1                                                              __init__.py:628
           DEBUG    Adding current directory to system path                                              __init__.py:71
           DEBUG    Finding namespace plugins                                                              _utils.py:97
           DEBUG    Loading plugin `m.abbr`                                                               _utils.py:105
           DEBUG    Loading plugin `m.alias`                                                              _utils.py:105
           DEBUG    Loading plugin `m.code`                                                               _utils.py:105
           DEBUG    Loading plugin `m.components`                                                         _utils.py:105
           DEBUG    Loading plugin `m.dox`                                                                _utils.py:105
           DEBUG    Loading plugin `m.dot`                                                                _utils.py:105
           DEBUG    Loading plugin `m.filesize`                                                           _utils.py:105
           DEBUG    Loading plugin `m.gl`                                                                 _utils.py:105
           DEBUG    Loading plugin `m.gh`                                                                 _utils.py:105
           DEBUG    Loading plugin `m.htmlsanity`                                                         _utils.py:105
           DEBUG    Loading plugin `m.images`                                                             _utils.py:105
           DEBUG    Loading plugin `m.link`                                                               _utils.py:105
           DEBUG    Loading plugin `m.math`                                                               _utils.py:105
Warning: libgs not found
           DEBUG    Loading plugin `m.metadata`                                                           _utils.py:105
           DEBUG    Loading plugin `m.plots`                                                              _utils.py:105
           ERROR    Cannot load plugin `m.plots`                                                              log.py:94
                    No module named 'matplotlib'
           DEBUG    Loading plugin `m.sphinx`                                                             _utils.py:105
           DEBUG    Loading plugin `m.qr`                                                                 _utils.py:105
           ERROR    Cannot load plugin `m.qr`                                                                 log.py:94
                    No module named 'qrcode'
           DEBUG    Loading plugin `m.vk`                                                                 _utils.py:105
           DEBUG    Registering plugin `m.abbr`                                                          __init__.py:78
           DEBUG    Registering plugin `m.alias`                                                         __init__.py:78
           DEBUG    Registering plugin `m.code`                                                          __init__.py:78
           DEBUG    Registering plugin `m.components`                                                    __init__.py:78
           DEBUG    Registering plugin `m.dox`                                                           __init__.py:78
           DEBUG    Registering plugin `m.dot`                                                           __init__.py:78
           DEBUG    Registering plugin `m.filesize`                                                      __init__.py:78
           DEBUG    Registering plugin `m.gl`                                                            __init__.py:78
           DEBUG    Registering plugin `m.gh`                                                            __init__.py:78
           DEBUG    Registering plugin `m.htmlsanity`                                                    __init__.py:78
           DEBUG    Registering plugin `m.images`                                                        __init__.py:78
           DEBUG    Registering plugin `m.link`                                                          __init__.py:78
           DEBUG    Registering plugin `m.math`                                                          __init__.py:78
           DEBUG    Registering plugin `m.metadata`                                                      __init__.py:78
           DEBUG    Registering plugin `m.sphinx`                                                        __init__.py:78
           DEBUG    Registering plugin `m.vk`                                                            __init__.py:78
  --- AutoReload Mode: Monitoring `content`, `theme` and `settings` for changes. ---
Serving site at: http://127.0.0.1:8000 - Tap CTRL-C to stop
Warning: libgs not found
Cannot load plugin `m.plots`
No module named 'matplotlib'
Cannot load plugin `m.qr`
No module named 'qrcode'
Unable to find '/posts/why-study-dsa.rst', skipping url replacement.
Done: Processed 2 articles, 0 drafts, 0 hidden articles, 2 pages, 0 hidden pages and 0 draft pages in 0.56 seconds.
```
Notice the line that says 'Unable to find `/posts/why-study-dsa.rst, skipping URL replacement`. This references an old folder name, which I changed to `articles/` since like four days ago. I've made several significant changes to the content of the articles, but it doesn't reflect on the website. 
I believe all these issues are related. 

Deleting the output folder and regenerating the content doesn't fix it. The fix suggested by @lpirl also didn't work for me. Please help!
